File: python/tests/presence_test.py
"""Presence / gRPC (S-PR) acceptance scenarios, including the catching half of
the G-GRPC case (a partial stream must not be returned as complete).
"""

from __future__ import annotations

from harness.fixture import Fixture
from tests.helpers import client_for, seed_channel, seed_member, seed_user


def test_s_pr_01_unary_online_from_presence(fixture: Fixture) -> None:
    a = seed_user(fixture, "pra")
    b = seed_user(fixture, "prb")
    fixture.presence.set_online(b.id)
    response = client_for(fixture, a.id).get(f"/users/{b.id}/presence")
    response.expect(200)
    assert response.json()["status"] == "online"


def test_s_pr_02_unary_offline_without_heartbeat(fixture: Fixture) -> None:
    a = seed_user(fixture, "pra")
    c = seed_user(fixture, "prc")
    response = client_for(fixture, a.id).get(f"/users/{c.id}/presence")
    response.expect(200)
    assert response.json()["status"] == "offline"


def test_s_pr_03_channel_presence_complete_stream(fixture: Fixture) -> None:
    owner = seed_user(fixture, "prowner")
    members = [seed_user(fixture, f"prm{i}") for i in range(4)]
    channel_id = seed_channel(fixture, owner, "general", private=True)
    for member in members:
        seed_member(fixture, owner, channel_id, member)
    # 2 of the 5 online.
    fixture.presence.set_online(owner.id)
    fixture.presence.set_online(members[0].id)

    response = client_for(fixture, owner.id).get(f"/channels/{channel_id}/presence")
    response.expect(200)
    entries = response.json()["members"]
    assert len(entries) == 5  # stream consumed to completion
    online = {e["userId"] for e in entries if e["status"] == "online"}
    assert online == {owner.id, members[0].id}


def test_s_pr_04_grpc_partial_stream_is_502(fixture: Fixture) -> None:
    """[G-GRPC] harness arms "stream fails after 2 messages" → channel presence →
    502; response contains NO partial list."""
    owner = seed_user(fixture, "prowner")
    members = [seed_user(fixture, f"prm{i}") for i in range(4)]
    channel_id = seed_channel(fixture, owner, "general", private=True)
    for member in members:
        seed_member(fixture, owner, channel_id, member)
    fixture.presence.fail_stream_after(2)

    response = client_for(fixture, owner.id).get(f"/channels/{channel_id}/presence")
    response.expect(502).expect_code("presence:incomplete")
    assert "members" not in response.json()


def test_s_pr_05_non_member_presence(fixture: Fixture) -> None:
    owner = seed_user(fixture, "prowner")
    outsider = seed_user(fixture, "prout")
    public_id = seed_channel(fixture, owner, "pub", private=False)
    private_id = seed_channel(fixture, owner, "priv", private=True)
    client_for(fixture, outsider.id).get(f"/channels/{public_id}/presence").expect(403)
    client_for(fixture, outsider.id).get(f"/channels/{private_id}/presence").expect(404)

File: go/tests/presence_test.go
package relaytest

import (
	"encoding/json"
	"net/http"
	"testing"

	"github.com/drim-dev/verifying-agent-code/go/src/relay/app"
	"github.com/drim-dev/verifying-agent-code/go/src/relay/domain"
)

func TestS_PR_01_unary_online_from_presence(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	bob := seedUser(t, "bob")
	fixture.Presence.SetOnline(bob.ID)
	var body struct {
		Status string `json:"status"`
	}
	newClient(t, ada.ID).get("/users/" + bob.ID + "/presence").expect(http.StatusOK).decode(&body)
	if body.Status != "online" {
		t.Fatalf("expected online when the presence service reports online, got %q", body.Status)
	}
}

func TestS_PR_02_no_heartbeat_is_offline(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	cleo := seedUser(t, "cleo")
	var body struct {
		Status string `json:"status"`
	}
	newClient(t, ada.ID).get("/users/" + cleo.ID + "/presence").expect(http.StatusOK).decode(&body)
	if body.Status != "offline" {
		t.Fatalf("expected offline without heartbeat, got %q", body.Status)
	}
}

func TestS_PR_03_channel_presence_complete(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	ch := seedChannel(t, ada, "team", false)
	members := []domain.User{ada}
	for _, h := range []string{"b", "c", "d", "e"} {
		u := seedUser(t, "user_"+h)
		seedMember(t, ada, ch, u)
		members = append(members, u)
	}
	fixture.Presence.SetOnline(members[1].ID)
	fixture.Presence.SetOnline(members[2].ID)

	var body struct {
		Members []struct {
			UserID string `json:"userId"`
			Status string `json:"status"`
		} `json:"members"`
	}
	newClient(t, ada.ID).get("/channels/" + ch + "/presence").expect(http.StatusOK).decode(&body)
	if len(body.Members) != 5 {
		t.Fatalf("expected 5 presence entries, got %d", len(body.Members))
	}
	online := 0
	for _, m := range body.Members {
		if m.Status == "online" {
			online++
		}
	}
	if online != 2 {
		t.Fatalf("expected exactly 2 online, got %d", online)
	}
}

// ---- G-GRPC: a mid-stream error must surface as 502, never a partial list ----

func TestS_PR_04_partial_stream_is_502(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	ch := seedChannel(t, ada, "team", false)
	for _, h := range []string{"b", "c", "d"} {
		seedMember(t, ada, ch, seedUser(t, "user_"+h))
	}
	assertPartialStreamRejected(t, fixture.BaseURL(), ada, ch)
}

func TestS_PR_04_naive_swallows_partial_is_caught(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	ch := seedChannel(t, ada, "team", false)
	for _, h := range []string{"b", "c", "d"} {
		seedMember(t, ada, ch, seedUser(t, "user_"+h))
	}
	naive := fixture.NaiveApp(func(d *app.Deps) { d.Presence = naivePresenceClient{correct: d.Presence} })
	defer naive.Close()
	expectCatchToFail(t, "G-GRPC", func(tb testing.TB) {
		assertPartialStreamRejected(tb, naive.BaseURL(), ada, ch)
	})
}

func assertPartialStreamRejected(tb testing.TB, baseURL string, caller domain.User, channelID string) {
	tb.Helper()
	fixture.Presence.FailStreamAfter(2)
	resp := clientAt(tb, baseURL, caller.ID).get("/channels/" + channelID + "/presence")
	resp.expect(http.StatusBadGateway).expectCode("presence:incomplete")
	var body struct {
		Members []json.RawMessage `json:"members"`
	}
	if len(resp.body) > 0 && resp.body[0] == '{' {
		_ = json.Unmarshal(resp.body, &body)
	}
	if len(body.Members) != 0 {
		tb.Fatalf("a 502 must carry NO partial member list, got %d entries", len(body.Members))
	}
}

func TestS_PR_05_channel_presence_visibility(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	bob := seedUser(t, "bob")
	pub := seedChannel(t, ada, "public", false)
	priv := seedChannel(t, ada, "secret", true)
	newClient(t, bob.ID).get("/channels/" + pub + "/presence").
		expect(http.StatusForbidden).expectCode("channel:membership_required")
	newClient(t, bob.ID).get("/channels/" + priv + "/presence").
		expect(http.StatusNotFound).expectCode("channel:not_found")
}
